File: scripts/rustdoc_coverage.py
#!/usr/bin/env python3
"""
Script to analyze rustdoc coverage in pdftract-core.

Measures:
- Total public items (pub fn, pub struct, pub enum, pub trait, pub type)
- Public items with documentation
- Public items with worked examples (```rust blocks)
"""
import subprocess
import re
from pathlib import Path
from collections import defaultdict
from dataclasses import dataclass
from typing import Dict, List

@dataclass
class ModuleStats:
    total: int = 0
    with_doc: int = 0
    with_example: int = 0
    items: List[str] = None

    def __post_init__(self):
        if self.items is None:
            self.items = []

def run_rg(pattern: str, path: Path) -> str:
    """Run ripgrep and return output."""
    result = subprocess.run(
        ["rg", pattern, str(path), "-n", "-A", "10", "--type", "rust"],
        capture_output=True,
        text=True,
        cwd="/home/coding/pdftract"
    )
    return result.stdout

def analyze_module(module_path: Path) -> ModuleStats:
    """Analyze a single module file for rustdoc coverage."""
    stats = ModuleStats()

    content = module_path.read_text()
    lines = content.split("\n")

    # Track public items
    for i, line in enumerate(lines):
        # Look for pub items
        for pattern in [
            r"pub\s+fn\s+(\w+)",
            r"pub\s+struct\s+(\w+)",
            r"pub\s+enum\s+(\w+)",
            r"pub\s+trait\s+(\w+)",
            r"pub\s+type\s+(\w+)",
            r"pub\s+mod\s+(\w+)",
        ]:
            match = re.search(pattern, line)
            if match:
                item_name = match.group(1)
                stats.total += 1
                stats.items.append(f"{line.strip()}:{i+1}")

                # Check for documentation above
                has_doc = False
                has_example = False

                # Look back up to 20 lines for doc comments
                for j in range(max(0, i - 20), i):
                    prev_line = lines[j].strip()
                    if prev_line.startswith("///") or prev_line.startswith("//!"):
                        has_doc = True
                        # Check for example within doc
                        if "```rust" in prev_line or "```rust,no_run" in prev_line or "```ignore" in prev_line:
                            has_example = True
                        # Also check a few lines after the doc start
                        for k in range(j+1, min(j+10, i)):
                            if "```rust" in lines[k]:
                                has_example = True
                    elif not prev_line.startswith("//") and prev_line and not prev_line.startswith("#"):
                        # Stop if we hit something that's not a comment
                        if j < i - 1 and lines[j+1].strip().startswith("#"):
                            continue
                        if j < i - 2:
                            break

                if has_doc:
                    stats.with_doc += 1
                if has_example:
                    stats.with_example += 1

    return stats

def main():
    """Main analysis function."""
    src_dir = Path("/home/coding/pdftract/crates/pdftract-core/src")

    print(f"Analyzing rustdoc coverage for pdftract-core")
    print(f"=" * 60)

    total_stats = ModuleStats()
    module_stats: Dict[str, ModuleStats] = {}

    # Analyze each module
    for rs_file in sorted(src_dir.rglob("*.rs")):
        # Skip main.rs and test files
        if "tests" in str(rs_file) or rs_file.name == "main.rs":
            continue

        # Get module name from path
        rel_path = rs_file.relative_to(src_dir)
        if str(rel_path) == "lib.rs":
            continue

        module_name = str(rel_path).replace("/", "::").replace(".rs", "")
        stats = analyze_module(rs_file)

        if stats.total > 0:
            module_stats[module_name] = stats
            total_stats.total += stats.total
            total_stats.with_doc += stats.with_doc
            total_stats.with_example += stats.with_example

    # Print report
    print(f"\nOverall Coverage:")
    print(f"  Total public items: {total_stats.total}")
    print(f"  With documentation: {total_stats.with_doc} ({100*total_stats.with_doc/total_stats.total:.1f}%)")
    print(f"  With examples: {total_stats.with_example} ({100*total_stats.with_example/total_stats.total:.1f}%)")
    print()

    print(f"Top modules by public items:")
    sorted_modules = sorted(module_stats.items(), key=lambda x: x[1].total, reverse=True)[:15]
    for name, stats in sorted_modules:
        doc_pct = 100 * stats.with_doc / stats.total if stats.total > 0 else 0
        ex_pct = 100 * stats.with_example / stats.total if stats.total > 0 else 0
        print(f"  {name:50s} items:{stats.total:3d} docs:{doc_pct:5.1f}% examples:{ex_pct:5.1f}%")

if __name__ == "__main__":
    main()

File: scripts/rustdoc_coverage.rs
#!/usr/bin/env rust-script
//! Scan pdftract-core source for public API items with/without worked examples.

use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};
use syn::{Attribute, Item, ItemEnum, ItemFn, ItemStruct, ItemTrait, ItemMod, ItemType, Visibility};

#[derive(Debug, Default)]
struct ModuleStats {
    total_items: usize,
    with_examples: usize,
    missing_docs: usize,
    items: Vec<ItemInfo>,
}

#[derive(Debug)]
struct ItemInfo {
    name: String,
    kind: &'static str,
    has_example: bool,
    file: String,
    line: usize,
}

fn extract_examples_from_doc(attrs: &[Attribute]) -> bool {
    for attr in attrs {
        if let syn::Meta::NameValue(meta) = &attr.meta {
            if meta.path.is_ident("doc") {
                if let Ok(syn::Expr::Lit(expr_lit)) = &meta.value {
                    if let syn::Lit::Str(lit_str) = &expr_lit.lit {
                        let doc = lit_str.value();
                        // Check for ```rust code blocks (worked examples)
                        if doc.contains("```rust") || doc.contains("```no_run") || doc.contains("```ignore") {
                            return true;
                        }
                    }
                }
            }
        }
    }
    false
}

fn count_public_items_in_file(content: &str, file: &Path) -> Vec<ItemInfo> {
    let mut items = Vec::new();

    let file = file.to_path_buf();
    let syntax = match syn::parse_file(content) {
        Ok(s) => s,
        Err(e) => {
            eprintln!("Failed to parse {}: {}", file.display(), e);
            return items;
        }
    };

    for item in syntax.items {
        match item {
            Item::Fn(ItemFn { attrs, vis, sig, .. }) => {
                if matches!(vis, Visibility::Public(_)) {
                    let name = sig.ident.to_string();
                    let has_example = extract_examples_from_doc(&attrs);
                    items.push(ItemInfo {
                        name,
                        kind: "fn",
                        has_example,
                        file: file.display().to_string(),
                        line: attrs.first().map(|a| a.span().start().line).unwrap_or(0),
                    });
                }
            }
            Item::Struct(ItemStruct { attrs, vis, ident, .. }) => {
                if matches!(vis, Visibility::Public(_)) {
                    let name = ident.to_string();
                    let has_example = extract_examples_from_doc(&attrs);
                    items.push(ItemInfo {
                        name,
                        kind: "struct",
                        has_example,
                        file: file.display().to_string(),
                        line: attrs.first().map(|a| a.span().start().line).unwrap_or(0),
                    });
                }
            }
            Item::Enum(ItemEnum { attrs, vis, ident, .. }) => {
                if matches!(vis, Visibility::Public(_)) {
                    let name = ident.to_string();
                    let has_example = extract_examples_from_doc(&attrs);
                    items.push(ItemInfo {
                        name,
                        kind: "enum",
                        has_example,
                        file: file.display().to_string(),
                        line: attrs.first().map(|a| a.span().start().line).unwrap_or(0),
                    });
                }
            }
            Item::Trait(ItemTrait { attrs, vis, ident, .. }) => {
                if matches!(vis, Visibility::Public(_)) {
                    let name = ident.to_string();
                    let has_example = extract_examples_from_doc(&attrs);
                    items.push(ItemInfo {
                        name,
                        kind: "trait",
                        has_example,
                        file: file.display().to_string(),
                        line: attrs.first().map(|a| a.span().start().line).unwrap_or(0),
                    });
                }
            }
            Item::Type(ItemType { attrs, vis, ident, .. }) => {
                if matches!(vis, Visibility::Public(_)) {
                    let name = ident.to_string();
                    let has_example = extract_examples_from_doc(&attrs);
                    items.push(ItemInfo {
                        name,
                        kind: "type",
                        has_example,
                        file: file.display().to_string(),
                        line: attrs.first().map(|a| a.span().start().line).unwrap_or(0),
                    });
                }
            }
            Item::Mod(ItemMod { attrs, vis, ident, .. }) => {
                if matches!(vis, Visibility::Public(_)) {
                    let name = ident.to_string();
                    let has_example = extract_examples_from_doc(&attrs);
                    items.push(ItemInfo {
                        name,
                        kind: "mod",
                        has_example,
                        file: file.display().to_string(),
                        line: attrs.first().map(|a| a.span().start().line).unwrap_or(0),
                    });
                }
            }
            _ => {}
        }
    }

    items
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let core_src = Path::new("crates/pdftract-core/src");
    let mut module_stats: HashMap<String, ModuleStats> = HashMap::new();

    for entry in walkdir::WalkDir::new(core_src) {
        let entry = entry?;
        let path = entry.path();

        if path.extension().and_then(|s| s.to_str()) != Some("rs") {
            continue;
        }

        let content = fs::read_to_string(path)?;
        let module_name = path
            .strip_prefix(core_src)
            .ok()
            .and_then(|p| p.parent())
            .and_then(|p| p.file_name())
            .and_then(|n| n.to_str())
            .unwrap_or("lib")
            .to_string();

        let items = count_public_items_in_file(&content, path);

        for item in items {
            let stats = module_stats
                .entry(module_name.clone())
                .or_insert_with(ModuleStats::default);
            stats.total_items += 1;
            if item.has_example {
                stats.with_examples += 1;
            }
            stats.items.push(item);
        }
    }

    let mut total_items = 0;
    let mut total_with_examples = 0;

    println!("\n=== Rustdoc Coverage Report for pdftract-core ===\n");

    for (module, stats) in module_stats.iter() {
        let coverage = if stats.total_items > 0 {
            (stats.with_examples as f64 / stats.total_items as f64) * 100.0
        } else {
            0.0
        };
        println!(
            "{}: {}/{} items with examples ({:.1}%)",
            module, stats.with_examples, stats.total_items, coverage
        );
        total_items += stats.total_items;
        total_with_examples += stats.with_examples;
    }

    let overall_coverage = if total_items > 0 {
        (total_with_examples as f64 / total_items as f64) * 100.0
    } else {
        0.0
    };

    println!(
        "\nOverall: {}/{} items with examples ({:.1}%)",
        total_with_examples, total_items, overall_coverage
    );

    if overall_coverage < 80.0 {
        println!("\n⚠️  Coverage is below 80% target");
    } else {
        println!("\n✅ Coverage meets 80%+ target");
    }

    // List items without examples (limited output)
    println!("\n=== Items without examples (first 20 per module) ===\n");
    for (module, stats) in module_stats.iter() {
        let without_examples: Vec<_> = stats
            .items
            .iter()
            .filter(|i| !i.has_example)
            .take(20)
            .collect();
        if !without_examples.is_empty() {
            println!("{}:", module);
            for item in without_examples {
                println!("  - {} ({}) at {}:{}", item.name, item.kind, item.file, item.line);
            }
            println!();
        }
    }

    Ok(())
}
